### DISCUSSION TITLE
Page image not being rendered during rewrite operations

### DISCUSSION THREAD

**Discussion by djoltes** [pymupdf/PyMuPDF#2557] [Looking for help]
I'm continuing development of a redaction solution, and recently noticed that in some cases pages are rendered with the redactions in place, but without the actual text backing the page. There seems to be no pattern: in some cases even a 1-page document is affected, while in others I'll see a mix of correctly rendered pages along with others where all that's visible are the redaction blocks.

![image](https://github.com/pymupdf/PyMuPDF/assets/11085091/79733618-a98b-47b9-a550-d10db7e6a75f)

No idea why this started happening recently...any ideas appreciated. I've been doing page-by-page intermediate saves of documents, which has never caused a problem. I may look at that as a possible cause.

**Comment by JorjMcKie** [comment]
Hard to tell without seeing the code.
Storing stuff to an existing page may go wrong when there are non-localized geometry changes as explained in ["Misplaced Item Insertions on PDF Pages"](https://pymupdf.readthedocs.io/en/latest/recipes-common-issues-and-their-solutions.html).
Try `page.clean_contents()` immediately after loading a page.

**Reply by djoltes** [reply]
Let me paste in what I hope are the salient bits of code. Effectively, what I need to do (at least for now) is to iterate over pages as usual, redacting each one, and since we've discovered a lot of the 'wild' documents we're dealing with have really poor OCR we're trying to use Textract to rewrite it. The idea is that at least we're only dealing with one OCR engine at that point (I may revert to Tesseract, but it has other issues). 

So...

```
for page in doc:
[...]
page.wrap_contents()
        
# Next, get an image of the page so we can determine its size in pixels. We need this 
# in order to place extracted text at the right coordinates during the next phases, 
# as Textract returns its results with a series of relativistic coordinates.
pix = page.get_pixmap()
page_jpg = pix.tobytes(output = 'jpg')

# We need to know the size of the page in pixels, so we have to call cv2 to retrieve this
img = np.asarray(bytearray(page_jpg), dtype="uint8")
img = cv2.imdecode(img, 0)
iHeight, iWidth = img.shape[:2]

[ call Textract to get the dictionary result, ocrDict...]

for item in ocrDict["Blocks"]:
                if item["BlockType"] == "LINE" or item["BlockType"] == "WORD":   
                    ocrText = item["Text"]
                    ocrConf = item["Confidence"]
                    if debug:
                        print("OCR Text: {a} Score: {b}%".format(a = ocrText, b = ocrConf))
                    # increment our OCR counters
                    pageTotalOCRevents += 1
                    pageTotalOCRscore += ocrConf
                    
                    # Now start computing the size of our rectangle annotation based on the bounding box data
                    geo = item["Geometry"]
                    box = geo["BoundingBox"]
                    x0 = box["Left"] * iWidth # left side
                    y0 = box["Top"] * iHeight
                    height = box["Height"] * iHeight 
                    width = box["Width"] * iWidth
                    x1 = x0 + width # computed width
                    y1 = y0 + height # computed height
        
                    matrix = fitz.Rect(0, 0, 1, 1).torect(page.rect)
                    ocrRect = fitz.Rect(x0, y0, x1, y1)
                    bbox = ocrRect * matrix
        
                    textLen = font.text_length(ocrText, fontsize=1)
                    fontSize = ocrRect.width / textLen
    
                    page.insert_text(ocrRect.bl, 
                                     ocrText,
                                     fontsize = fontSize,
                                     fontname = "helv",
                                     render_mode = 3)

[...now that we've OCR'd again we can call Comprehend and do redactions....]

if pageText:
            piiFromText = compr.detect_pii_entities(Text = pageText, LanguageCode = 'en')   

            for NER in piiFromText['Entities']:
                if NER['Type'] in ENT_TYPE_TO_REDACT: # passed test 1
                    if NER['Score'] > RED_THRESH:
                        startPoint = NER['BeginOffset']
                        endPoint = NER['EndOffset']
                        itemToRedact = pageText[startPoint:endPoint]
                        print("PII entity: {a}. Score: {b}".format(a = itemToRedact, b = NER['Score']))
                        logging.info("PII entity: {a}. Score: {b}".format(a = itemToRedact, b = NER['Score']))
                        pageTotalPIIevents += 1 
                        pageTotalPIIscore += NER['Score']
                        if NER['Score'] > pageMaxPIIscore:
                            pageMaxPIIscore = NER['Score']
                        if NER['Score'] < pageMinPIIscore:
                            pageMinPIIscore = NER['Score']
                        if itemToRedact not in piiList:
                            piiList.append(itemToRedact)
                    else: # we rejected the redaction due to a lower score, but need to note it
                        startPoint = NER['BeginOffset']
                        endPoint = NER['EndOffset']
                        itemToRedact = pageText[startPoint:endPoint]
                        print("Rejected PII entity: {a}. Score: {b}".format(a = itemToRedact, b = NER['Score']))
                        logging.info("Rejected PII entity: {a}. Score: {b}".format(a = itemToRedact, b = NER['Score']))
                        pageTotalPIIrejects += 1
                        pageTotalPIIrejscore += NER['Score']
                        if NER['Score'] < pageMinPIIscore:
                            pageMinPIIscore = NER['Score']
       
                
            if piiList is not None:
                for data in piiList:
                    # print("email: {a}".format(a=data))
                    areas = page.search_for(data)      
                    # draw outline over sensitive data areas based on coordinates on the page
                    [page.add_redact_annot(area, fill = (0, 0, 0), text=REDACT_STR) for area in areas]                
                        
[... do some more annotations based on regex results...]

        # apply the redactions
        try:
            # page.apply_redactions()
            page.apply_redactions(images = PDF_REDACT_IMAGE_NONE)
```
One additional problem is that, since some pages already have OCR text on them, I tried applying a suggestion I found on another thread about removing all existing text with the following, but it seems not to work -- so in those cases I get doubled text.

```
page.add_redact_annot(page.rect, fill = None, text="", text_color = None)
page.apply_redactions(images = fitz.PDF_REDACT_IMAGE_NONE)
```

I'm half considering taking a different tack altogether -- using the page image (page_jpg) as input to the various AWS services, redacting the jpg itself, then replacing it on the page while stripping out all the text. That should have a file with fully redacted images, but with no OCR layer at all. Not sure if it would work, though.

**Reply by JorjMcKie** [reply]
Comments to your code:
* `pix = page.get_pixmap()` - you may want to consider a DPI value for improved recognition; e.g. dpi=150 or higher. This will increase the confidence value.
* The Pixmap contains its width and height and resolution as properties: `pix.width`, `pix.height`, `pix.xres`, `pix.yres`. You don't need a separate package like opencv.

**Reply by djoltes** [reply]
Ah, I didn't know about the newer dpi option for the pixmap. I'll give that a try to see if it helps with resolution or redaction placement with the various AWS services. I read your comments in the other thread about the problems of reconstructing text via OCR and appreciate that fact -- this is not an easy thing to do, and as we all know PDF was originally intended as an output format rather than something to be mined for data. I've recently been doing data extraction from the same sort of 'wild' PDFs, and have had to resort to a lot of parsing tricks to get reasonably high levels of accuracy there as well. 

My bigger issue at present, though, is this problem with page images apparently not being rendered randomly.  Yesterday I created a test harness to just process one small file (OCR only, no redaction).

```
import fitz, logging
import re, glob, os, io, sys, cv2

# Set error display during mupdf ops to true so we can capture them in the error log
fitz.TOOLS.mupdf_display_errors(True)

from ocrmypdf.exceptions import BadArgsError
import boto3 # for Comprehend and Textract

REGION = 'us-east-1'
PG_OFFSET = 1
font = fitz.Font("helv") # create a fitz font instance

SEKRIT = 'redacted!'
PUBLIC = 'redacted!


def ocr_page_textract(page, pageLoc):
    fitz.TOOLS.mupdf_display_warnings(True)
    print("OCRing page")
    src = page.parent  # the page's document
    doc = fitz.open()  # make temporary 1-pager
    doc.insert_pdf(src, from_page=pageLoc, to_page=pageLoc)
    pdfbytes = doc.tobytes()
    inbytes = io.BytesIO(pdfbytes)  # transform to BytesIO object
    outbytes = io.BytesIO()
    textract_client = boto3.client('textract', 
        region_name = REGION, 
        aws_secret_access_key = SEKRIT, 
        aws_access_key_id = PUBLIC) 
    try:
        logging.getLogger().setLevel(logging.INFO)
        outbytes = textract_client.detect_document_text(Document={'Bytes': pdfbytes})
        
    except BadArgsError as e:
        print("File may be empty or corrupted") 
        print("Error: {a}".format(a = e))
        raise
    except TypeError as e:
        print("Invalid data in OCR output")
        print("Error: {a}".format(a = e))
        print(sys.exc_info()[0].__name__, os.path.basename(sys.exc_info()[2].tb_frame.f_code.co_filename), sys.exc_info()[2].tb_lineno)
    except:
        print(sys.exc_info()[0].__name__, os.path.basename(sys.exc_info()[2].tb_frame.f_code.co_filename), sys.exc_info()[2].tb_lineno)
   
    if outbytes is not None:
        # return t_text
        return outbytes
    else:
        return None        
    ## main logic
    
textract_client = boto3.client('textract', 
    region_name = REGION, 
    aws_secret_access_key = SEKRIT, 
    aws_access_key_id = PUBLIC)

# Open the file
f = "c:/analytics/pdf_samples/Redaction_test.pdf"
baseFname = os.path.basename(f).split('.')[0]
newFile = "c:/analytics/pdf_samples/{a}_redacted.pdf".format(a=baseFname)
print("output will be: {a}".format(a = newFile))

doc = fitz.open(f)   
print("opened file {a}".format(a = f))
# start iterating through pages
pageNo = 1
totalPages = 0
pageCount = len(doc)
fileTotalPages = 0

for page in doc:
    print("Page #: {a}".format(a = pageNo))
    logging.info("Page #: {a}".format(a = pageNo))
    fileTotalPages += 1 # increment page count
    myText = str(page.get_text().encode("utf8"))
    myText = re.sub("\n", "", myText)
    mytext = myText.strip()
    
    page.wrap_contents()
    
    # Next, get an image of the page so we can determine its size in pixels. We need this 
    # in order to place extracted text at the right coordinates during the next phases, 
    # as Textract returns its results with a series of relativistic coordinates.
    pix = page.get_pixmap()
    page_jpg = pix.tobytes(output = 'jpg')
    
    # We need to know the size of the page in pixels, so we have to call cv2 to retrieve this
    img = np.asarray(bytearray(page_jpg), dtype="uint8")
    img = cv2.imdecode(img, 0)
    iHeight, iWidth = img.shape[:2]
    print("Image size. Width: {a} Height: {b}".format(a = iWidth, b = iHeight))
    
    if True:
        print("OCR!")
        ocrDict = ocr_page_textract(page, page.number + PG_OFFSET)
        # page.add_redact_annot(page.rect, fill = None, text="", text_color = None)
        # page.apply_redactions(images = fitz.PDF_REDACT_IMAGE_NONE)
        
        for item in ocrDict["Blocks"]:
            if item["BlockType"] == "LINE" or item["BlockType"] == "WORD":   
                ocrText = item["Text"]
                ocrConf = item["Confidence"]
                
                # Now start computing the size of our rectangle annotation based on the bounding box data
                geo = item["Geometry"]
                box = geo["BoundingBox"]
                x0 = box["Left"] * iWidth # left side
                y0 = box["Top"] * iHeight
                height = box["Height"] * iHeight 
                width = box["Width"] * iWidth
                x1 = x0 + width # computed width
                y1 = y0 + height # computed height
    
                matrix = fitz.Rect(0, 0, 1, 1).torect(page.rect)
                ocrRect = fitz.Rect(x0, y0, x1, y1)
                bbox = ocrRect * matrix  
                textLen = font.text_length(ocrText, fontsize=1)
                fontSize = ocrRect.width / textLen    
                page.insert_text(ocrRect.bl, 
                                 ocrText,
                                 fontsize = fontSize,
                                 fontname = "helv",
                                 render_mode = 3)               
                # page.draw_rect(ocrRect, color = (.25, 1, 0.25))    
    if pageNo < pageCount:
        doc.save(newFile, garbage = 3, deflate=True)
        print("Intermediate save...")
        # Now move to the next page
        pageNo += 1
print("Saving final document")
doc.save(newFile, garbage = 3, deflate=True)
```
If I run this as-is (with the two lines noted below commented out) the text appears on the output file's page with no problem, but when you select all text and copy/paste into another file it's obvious it's just been duplicated by the OCR process. 

```
# page.add_redact_annot(page.rect, fill = None, text="", text_color = None)
# page.apply_redactions(images = fitz.PDF_REDACT_IMAGE_NONE)
```

If I uncomment those lines, I get a file that's just a white page with only the OCR text layer (also duplicated). The visible text is gone.

That said, in my main code it doesn't matter if I try the full-rect annotation or not...certain pages come out with just the redaction blocks showing, while others are clean.

The two blockers in my current effort are: (a) why am I getting random pages with no visible text, and (b) how can I remove any existing text from the page prior to running Textract against it?

**Reply by djoltes** [reply]
_Update_: I tried using the dpi = xxx option for the pixmap. I'm sure this helps with detection (it seemed to return better identifications) but it also has the effect of confusing placement on the page -- the bboxes were wildly off. I'd guess this is because the image dimensions (height/width) and returned offsets increase as resolution goes up, while the 'native' page image is still at default.  

What's the default dpi value for get_pixmap()? Maybe the resulting coordinates can be scaled down to map correctly to the native page image. When I use get_pixmap() at default for a typical page, the returned size is Width: 612 Height: 792 (+/-), but with dpi = 150 it comes back as Width: 1275 Height: 1650. That probably means the default is ~75dpi, so dividing the locations computed using height/width by 2 will probably result in bbox locations that are far more correct. Or am I blowing smoke?

Opinion?

**Reply by JorjMcKie** [reply]
For a given page rectangle, the pixmap rectangle (an `IRect`!) is computed as `(page.rect * dpi / 72).irect`. For example:
```python
In [27]: pix = page.get_pixmap(dpi=150)
In [28]: pix
Out[28]: Pixmap(DeviceRGB, IRect(0, 0, 1240, 1755), 0)
In [29]: (page.rect * 150 / 72).irect
Out[29]: IRect(0, 0, 1240, 1755)
```
You can transform between page rectangle and pixmap rectangle coordinates by computing the corresponding matrices:
```python
In [30]: pix_rect = pix.irect
In [31]: page_rect = page.rect
In [32]: page2pix = page_rect.torect(pix_rect)
In [33]: # a page rectangle (100, 100, 300, 200) to pix coordinates:
In [34]: fitz.Rect(100, 100, 300, 200) * page2pix
Out[34]: Rect(208.4033660888672, 208.4322967529297, 625.2101440429688, 416.8645935058594)
```
Similarly, pixmap (image) coordinates can be transformed back to page coords by multiplying with `~page2pix` the inverse matrix.

**Reply by djoltes** [reply]
Got it , thanks.  Should be easy to convert the dimensions back then.

The bigger issue is the original query though -- any idea how or why I'm getting these random "redaction only" pages?

**Reply by JorjMcKie** [reply]
> The bigger issue is the original query though -- any idea how or why I'm getting these random "redaction only" pages?

If you a full page redact annot and erase (all) the text under it (but ignoring the images! => REDACT_IMAGES_NONE), then you should get rid of previous text - hidden or not - while keeping the image with the embedded text for later OCRing.
No idea how this would not work!
Have an example page for testing this?

**Reply by djoltes** [reply]
Sure, here's a one-pager with some fake PII in it -- original and redacted versions.  The original file was created with Word and exported to PDF. The weird thing is that there seems to be no rhyme or reason for which pages exhibit this.

Oh, and at this point the code does not strip the initial text from the full page rect. 
[Redaction_test.pdf](https://github.com/pymupdf/PyMuPDF/files/12197390/Redaction_test.pdf)
[Redaction_test_redacted.pdf](https://github.com/pymupdf/PyMuPDF/files/12197393/Redaction_test_redacted.pdf)

**Reply by JorjMcKie** [reply]
Here is the explanation:
You must clean the page apearance commands before inserting the redactions!
The PDF creator (Word) might have left behind a mess of "geometry changes" of all sorts, causing your redact rectangle(s) to be potentially anything - except what you are expecting. The following does the job and leaves behind an empty page (except for two underlines of the links - which is inevitable:
```python
doc=fitz.open("Redaction_test.pdf")
page=doc[0]
page.is_wrapped
False
page.clean_contents()
page.is_wrapped
True
page.add_redact_annot(page.rect)
'Redact' annotation on page 0 of Redaction_test.pdf
page.apply_redactions()
True
doc.ez_save("original-redacted.pdf")
```

**Reply by djoltes** [reply]
Hmm, that's odd because I've tried using clean_contents() (I've also tried wrap_contents()) and I still get the same output.  This block is the initial page handling code I'm now using, and it gives the same results. 

```
    for page in doc:
        # wrap_contents is needed for fixing alignment issues with rect boxes 
        page.clean_contents()
        print("Page #: {a}".format(a = pageNo))
        logging.info("Page #: {a}".format(a = pageNo))
        fileTotalPages += 1 # increment page count
        myText = str(page.get_text().encode("utf8"))
        myText = re.sub("\n", " ", myText)
```

**Reply by djoltes** [reply]
Ahhh, I may have found the problem. One of the tasks involves using Textract to handle signature annotations, and it turns out something in that code was triggering an 'illicit' redaction of all the other text. I think I have it locked, but I'll run more tests.

Thanks as always for the back-and-forth. If you're ever near Boston, remind me I owe you a few pints!

**Reply by JorjMcKie** [reply]
😂 accepted!
